File: gitdot-web/packages/api/src/endpoint/repository/list-repository-commits.ts
import { z } from "zod";
import { RepositoryCommitsResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const ListRepositoryCommitsRequest = z.object({
  ref_name: z.string().optional(),
  from: z.iso.datetime().optional(),
  to: z.iso.datetime().optional(),
});
export type ListRepositoryCommitsRequest = z.infer<
  typeof ListRepositoryCommitsRequest
>;

export const ListRepositoryCommitsResponse = RepositoryCommitsResource;
export type ListRepositoryCommitsResponse = z.infer<
  typeof ListRepositoryCommitsResponse
>;

export const ListRepositoryCommits = {
  path: "/repository/{owner}/{repo}/commits",
  method: "GET",
  request: ListRepositoryCommitsRequest,
  response: ListRepositoryCommitsResponse,
} as const satisfies Endpoint;
export type ListRepositoryCommits = typeof ListRepositoryCommits;

File: gitdot-api/src/endpoint/repository/list_repository_commits.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use super::default_ref;
use crate::{endpoint::Endpoint, resource::repository::RepositoryCommitsResource};

pub struct ListRepositoryCommits;

impl Endpoint for ListRepositoryCommits {
    const PATH: &'static str = "/repository/{owner}/{repo}/commits";
    const METHOD: http::Method = http::Method::GET;

    type Request = ListRepositoryCommitsRequest;
    type Response = ListRepositoryCommitsResponse;
}

#[derive(ApiRequest, Debug, Serialize, Deserialize)]
pub struct ListRepositoryCommitsRequest {
    #[serde(default = "default_ref")]
    pub ref_name: String,
    pub from: Option<DateTime<Utc>>,
    pub to: Option<DateTime<Utc>>,
}

pub type ListRepositoryCommitsResponse = RepositoryCommitsResource;
